### ISSUE TITLE
`from-image`: can't specify silo image by name

### ISSUE THREAD

**Issue by david-crespo** [oxidecomputer/oxide.rs#411]
The code assumes that an image specified by name is a project image.

https://github.com/oxidecomputer/oxide.rs/blob/77a04019aa1ac794b28b2bfa158eb318b2d5a0fc/cli/src/cmd_instance.rs#L247-L251

I see two ways around this, neither great. I think I prefer the second.

* Add a way of indicating when you want a silo image by name
  * Instead of `--image` maybe have mutually exclusive `--project-image` and `--silo-image` flags?
  * I considered `--image` + `--image-project` but a) that project will always be the same as the `--project` project when specified, b) confusing name
* Avoid the problem by only allowing specifying an image by ID, which is pretty easily conveyed by the help text and the argument name. You could call it `--image-id`, that would be pretty unambiguous. That requires the user to use `oxide image view` view get the ID first.

Of course, by taking out half of what this command does for you, option 2 pushes us more in the direction of getting rid of it altogether.

> Yeah... I'm even more of the opinion that hand-rolled subcommands such as this one may be a step in the wrong direction.

https://github.com/oxidecomputer/oxide.rs/issues/373#issuecomment-1734554026

**Comment by rcgoodfellow** [comment]
Forcing the use of UUIDs would make the CLI very tedious to use. If something has a name, that name should be usable from the CLI.

The `from-image` CLI command is an example of providing an efficient user experience for a common task by composing API calls. It allows the API to take the approach of providing non-overlapping primitive functions while still providing a good user experience. If we can come up with a way to automate the generation of command composites for the CLI in some sort of declarative way that minimizes maintenance overhead, that would be excellent. But I think for the sake of usability, we need to have a way to provide these sorts of composite functions in the CLI.

**Comment by david-crespo** [comment]
I don't think we can generate it. In such cases the push would have to be to make a single API call do the thing we want. In this case that would probably mean making the API accept names [in addition to IDs](https://github.com/oxidecomputer/omicron/blob/36e047091a0793ae1d3a5f9bb176a2f7de137f5a/nexus/types/src/external_api/params.rs#L1138-L1154) for the disk create part of the call, that way we could skip the separate image fetch to resolve the name. That would still leave us with the problem of how to represent that stuff as command-line flags. I don't think we have a nice scheme yet for turning nested JSON objects into a flat list of flags. Seems like a minefield with few good options.

I am on board with the broader argument for convenience functions — this is basically the argument for the entire web console, and what `from-image` does is a lot like what the web console does on the instance create form. I guess it just stands out more in this repo because it's one of the few hand-rolled bits, whereas the console is mostly hand-rolled bits.

Another direction for this command would be the opposite of eliminating it — go maximalist, keeping the existing set of flags for use in scripts (adding something like `--image-project` or whatever to make silo images work by name), but also provide an interactive mode that resembles the web form:

![CLI interactive CLI select](https://raw.githubusercontent.com/mikaelmello/inquire/b79036ebc29ecfd6f7a1a039485d5b7b4558df0b/assets/select.gif)

https://github.com/mikaelmello/inquire?tab=readme-ov-file#select

I wonder if we could do this in a general way, allowing all commands to have a flag that says "give me interactive mode for any args I haven't provided".

**Comment by karencfv** [comment]
Unfortunately, interactive mode can be a minefield when using a CLI in scripts or any form of automation. Even if there's a way to turn it off it can trip people, and because the command isn't exiting it can be annoying to debug. This was something that was a bit of a problem with the previous CLI.

Would something like `--image` + `--image-scope={silo,project}` work?

**Comment by david-crespo** [comment]
Yeah, I think that would be fine. Could probably default to silo if it’s not present.
